spec/formatting.md

```diff
@@ -10,14 +10,18 @@ the successor to ICU MessageFormat, henceforth called ICU MessageFormat 1.0.
 
 ## Variable Resolution
 
-To resolve the value of a Variable,
-its Name is used to identify either a local variable,
+To resolve the value of a _variable_,
+its _name_ is used to identify either a local variable,
 or a variable defined elsewhere.
-If a local variable and an externally defined one use the same name,
-the local variable takes precedence.
 
-It is an error for a local variable definition to
-refer to a local variable that's defined after it in the message.
+If more than one _declaration_ binds a value to the same _name_,
+or if an externally defined variable and a _declaration_ use the same _name_,
+the resolved value of the most recent _declaration_ preceding the _variable_ is used.
```

Comment:
I think the function purity / lack of side effects ought to be addressed with respect to function resolution, rather than here.

---

The proposed "is processed sequentially" statement is a bit problematic, as it may be interpreted to require and enforce an eager processing model. Consider this message:

```
let $foo = {:really-heavy-operation}
match {$bar}
when 0 {Nothing here}
when * {Have a {$foo}}
```

In this case, we should make sure that in the `bar == 0` case an implementation is not required to determine and then throw away the value of `$foo`.

Reply:
Note that there is already an open issue about whether to require eager or lazy evaluation of local variables:

https://github.com/unicode-org/message-format-wg/issues/299

Thta said, I don't think either version of the wording commits to either eager or lazy evaluation. It depends on the interpretation of the word "processed". "Processing" could mean evaluating (resolving?) the right-hand side of the declaration and binding the left-hand side to the resulting value, or it could mean binding the left-hand side to a thunk representing the right-hand side.

Reply:
@eemeli I don't think we have to say anything about early or late binding of variables: we should not require early binding but we should not require late binding either. It's an implementation detail. What is important is that declarations behave as if they were evaluated in order.

I think we can let the text go in as-is and revisit.